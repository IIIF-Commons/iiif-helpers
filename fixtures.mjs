// Add thumbnail fixtures here.

// @type any
export const thumbnailFixtures = [
  {
    label: 'Canvas with no thumbnail property - ImageService',
    url: 'https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/no_thum_prop/v2/image_service2/manifest.json',
    description: 'This is a manifest where there is no thumbnail property but the thumbnail can be generated from the image service.',
  },
  {
    label: 'Canvas with no thumbnail property - level0 image',
    url: 'https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/no_thum_prop/v2/level0/manifest.json',
    description: 'This is a manifest where there is no thumbnail property but the thumbnail can be generated from the level 0 image service.',
  },
  {
    label: 'Canvas with no thumbnail property - simple image',
    url: 'https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/no_thum_prop/v2/simple_image/manifest.json',
    description: 'This is a manifest where there is no thumbnail property. There is a single image painted onto the canvas'
  },
  {
    label: 'Canvases with thumbnail property which is an image URL',
    url: 'https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/thum_prop/v2/canvas_thum_url/manifest.json',
    description: 'Each canvas has a thumbnail property which is a URL to an image',
  },
  {
    label: 'Canvases with thumbnail property which is an image URL with dimensions',
    url: 'https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/thum_prop/v2/canvas_thum_dimensions/manifest.json',
    description: 'Each canvas has a thumbnail property which is a URL to an image. The thumbnail property includes a type but no image service'
  },
  {
    label: 'Canvases with thumbnail property with level0 image service (Most optimised option)',
    url: 'https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/thum_prop/v2/canvas_thum_level0/manifest.json',
    description: 'This is an efficient way of advertising various sizes for a thumbnail and allows a producer to cache copies of thumbnails to make them fast to retrieve.',
  },
];
