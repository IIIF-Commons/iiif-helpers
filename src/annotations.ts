import { ChoiceBody, ImageService, InternationalString } from '@iiif/presentation-3';
import { ParsedSelector, SupportedTarget } from './annotation-targets';
import { ChoiceDescription, ComplexChoice, SingleChoice } from './painting-annotations';

// A wrapper for annotations that are "well-known"
interface ClassifiedAnnotation {
  id: string;
  type: 'Annotation';
  label?: InternationalString;
  summary?: InternationalString;
  motivation: string;
  target: SupportedTarget;
  body: SupportedBody;
}

interface AnnotationClassificationOptions {
  language?: string;
  choices?: string[];
  isTrusted?: boolean;
}

interface SimpleImage {
  id: string;
  format: string;
  label?: InternationalString;
  summary?: InternationalString;
  width?: number;
  height?: number;
  service?: ImageService;
}

type SupportedBody = {
  type: string;
  order: Array<'label' | 'summary' | 'image' | 'link' | 'internalLink' | 'content'>; // e.g. ['label', 'image', 'link', 'image']
  label?: InternationalString;
  summary?: InternationalString;
  content: InternationalString;
  image?: SimpleImage;
  images?: SimpleImage[];
  internalLink: {
    id: string;
    label?: InternationalString;
    type: 'Canvas';
    target: ParsedSelector;
  } | null;
  link: {
    id: string;
    label?: InternationalString;
    language?: string;
  } | null;
  contentList: InternationalString[];
  rawBody: any[];
  containsHtml: boolean;
  choice: ChoiceDescription | null;
};
