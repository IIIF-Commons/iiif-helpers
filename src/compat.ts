export type CompatVault = {
  get: import('@iiif/vault').Vault['get'];
  setMetaValue: import('@iiif/vault').Vault['setMetaValue'];
  getResourceMeta: import('@iiif/vault').Vault['getResourceMeta'];
};

const metaState: any = {};
export const compatVault: CompatVault = {
  get(nonRef: any) {
    return nonRef;
  },
  setMetaValue([id, meta, key], value) {
    const oldValue = compatVault.getResourceMeta(id, meta);
    const oldValueItem = oldValue ? oldValue[key] : undefined;
    const newValue = typeof value === 'function' ? (value as any)(oldValueItem) : value;
    metaState[id] = {
      ...(metaState[id] || {}),
      [meta]: {
        ...((metaState[id] || {})[meta] || {}),
        [key]: newValue,
      },
    };
  },
  getResourceMeta: ((resource: any, metaKey?: any) => {
    const resourceMeta = metaState[resource as any] as any;

    if (!resourceMeta) {
      return undefined;
    }
    if (!metaKey) {
      return resourceMeta;
    }

    return resourceMeta[metaKey];
  }) as any,
};
