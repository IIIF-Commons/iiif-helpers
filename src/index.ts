export * from './events';
export * from './styles';
export * from './thumbnail';
export * from './i18n';
