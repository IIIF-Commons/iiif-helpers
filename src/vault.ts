export * from './vault/vault';
export * from './vault/global-vault';
export * from './vault/types';
