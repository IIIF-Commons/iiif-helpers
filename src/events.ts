import type { Reference } from '@iiif/presentation-3';
import { compatVault, CompatVault } from './compat';

export function createEventsHelper(vault: CompatVault = compatVault) {
  return {
    addEventListener<T>(
      resource: Reference<any>,
      event: string,
      listener: (e: any, resource: T) => void,
      scope?: string[]
    ) {
      if (!resource) {
        return;
      }

      vault.setMetaValue<Array<{ callback: any; scope?: string[] }>>(
        [resource.id, 'eventManager', event],
        (registeredCallbacks) => {
          const callbacks = registeredCallbacks || [];
          for (const registered of callbacks) {
            if (registered.callback === listener) {
              // @todo check for scopes matching, very edge-case as scopes should be fixed.
              return callbacks;
            }
          }
          return [...callbacks, { callback: listener, scope }];
        }
      );

      return listener;
    },

    removeEventListener<T>(resource: Reference<any>, event: string, listener: (e: any, resource: T) => void) {
      if (!resource) {
        return;
      }
      vault.setMetaValue<Array<{ callback: () => void; scope?: string[] }>>(
        [resource.id, 'eventManager', event],
        (registeredCallbacks) => {
          return (registeredCallbacks || []).filter((registeredCallback) => registeredCallback.callback !== listener);
        }
      );
    },

    getListenersAsProps(resourceOrId: string | Reference<any>, scope?: string[]) {
      const resource = typeof resourceOrId === 'string' ? { id: resourceOrId } : resourceOrId;
      if (!resource || !resource.id) {
        return {};
      }
      const hooks = vault.getResourceMeta(resource.id, 'eventManager');
      const props: any = {};
      if (hooks && resource) {
        for (const hook of Object.keys(hooks)) {
          props[hook] = (e: any) => {
            const fullResource = vault.get<any>(resource);
            for (const { callback, scope: _scope } of hooks[hook] || []) {
              if (!_scope || (scope && _scope.indexOf(scope) !== -1)) {
                callback(e, fullResource);
              }
            }
          };
        }
      }
      return props;
    },
  };
}
