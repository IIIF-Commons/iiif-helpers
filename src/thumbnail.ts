import { ChoiceBody, CollectionItemSchemas, ContentResource, Reference } from '@iiif/presentation-3';
import {
  AnnotationNormalized,
  AnnotationPageNormalized,
  CanvasNormalized,
  CollectionNormalized,
  DescriptiveNormalized,
  ManifestNormalized,
} from '@iiif/presentation-3-normalized';
import {
  FixedSizeImage,
  FixedSizeImageService,
  getFixedSizeFromImage,
  ImageCandidate,
  ImageCandidateRequest,
  ImageServiceLoader,
  UnknownSizeImage,
  VariableSizeImage,
} from '@atlas-viewer/iiif-image-api';
import { compatVault, CompatVault } from './compat';

export const imageServiceLoader = new ImageServiceLoader();

let helper: ReturnType<typeof createThumbnailHelper> | null = null;
export function getThumbnail(
  input:
    | string
    | Reference<CollectionItemSchemas>
    | Reference<'Collection'>
    | Reference<'Manifest'>
    | Reference<'Canvas'>
    | Reference<'Annotation'>
    | Reference<'AnnotationPage'>
    | Reference<'ContentResource'>
    | CollectionNormalized
    | ManifestNormalized
    | CanvasNormalized
    | AnnotationNormalized
    | AnnotationPageNormalized
    | ContentResource
    | undefined,
  {
    vault = compatVault,
    dereference = false,
    ...options
  }: ImageCandidateRequest & { vault?: CompatVault; dereference?: boolean } = {}
) {
  helper = helper || createThumbnailHelper(vault);
  return helper.getBestThumbnailAtSize(input, options, dereference);
}

export function createThumbnailHelper(
  vault: CompatVault = compatVault,
  dependencies: { imageServiceLoader?: ImageServiceLoader } = {}
) {
  const loader = dependencies.imageServiceLoader || imageServiceLoader;

  async function getBestThumbnailAtSize(
    input:
      | string
      | Reference<CollectionItemSchemas>
      | Reference<'Collection'>
      | Reference<'Manifest'>
      | Reference<'Canvas'>
      | Reference<'Annotation'>
      | Reference<'AnnotationPage'>
      | Reference<'ContentResource'>
      | CollectionNormalized
      | ManifestNormalized
      | CanvasNormalized
      | AnnotationNormalized
      | AnnotationPageNormalized
      | ContentResource
      | undefined,
    request: ImageCandidateRequest,
    dereference?: boolean,
    candidates: Array<ImageCandidate> = [],
    dimensions?: { width: number; height: number }
  ): Promise<{
    best: null | undefined | FixedSizeImage | FixedSizeImageService | VariableSizeImage | UnknownSizeImage;
    fallback: Array<ImageCandidate>;
    log: string[];
  }> {
    const thumbnailNotFound = () => loader.getThumbnailFromResource(undefined as any, request, dereference, candidates);

    if (!input) {
      // We might have candidates already to pick from.
      return await loader.getThumbnailFromResource(undefined as any, request, dereference, candidates);
    }

    if (typeof input === 'string') {
      const fixed = getFixedSizeFromImage(input as any);
      if (fixed) {
        candidates.push(fixed);
      }

      return await loader.getThumbnailFromResource(undefined as any, request, dereference, candidates);
    }

    // Run through from ref, just in case.
    const fullInput:
      | string
      | ManifestNormalized
      | CollectionNormalized
      | CanvasNormalized
      | AnnotationNormalized
      | AnnotationPageNormalized
      | ContentResource
      | ChoiceBody
      | undefined = vault.get(input as any, { skipSelfReturn: false }) as any;

    if (typeof fullInput === 'string') {
      return { best: getFixedSizeFromImage(fullInput as any), fallback: [], log: [] };
    }

    if (!fullInput) {
      return await thumbnailNotFound();
    }

    const parseThumbnail = async (resource: DescriptiveNormalized) => {
      if (resource && resource.thumbnail && resource.thumbnail.length) {
        const thumbnail = vault.get(resource.thumbnail[0]);
        const potentialThumbnails = await loader.getImageCandidates(thumbnail as any, dereference);
        if (potentialThumbnails && potentialThumbnails.length) {
          candidates.push(...potentialThumbnails);
        }
      }
    };

    await parseThumbnail(fullInput as any);

    switch (fullInput.type) {
      case 'Annotation': {
        // Grab the body.
        const contentResources = Array.isArray(fullInput.body) ? fullInput.body : [fullInput.body];
        // @todo this could be configuration.
        const firstContentResources = vault.get(contentResources[0]);
        if (dimensions && !(firstContentResources as any).width) {
          (firstContentResources as any).width = dimensions.width;
          (firstContentResources as any).height = dimensions.height;
        }

        return await loader.getThumbnailFromResource(firstContentResources as any, request, dereference, candidates);
      }

      case 'Canvas': {
        const canvas = fullInput as CanvasNormalized;

        return getBestThumbnailAtSize(canvas.items[0], request, dereference, candidates, {
          width: canvas.width,
          height: canvas.height,
        });
      }

      // Unsupported for now.
      case 'AnnotationPage': {
        const annotationPage = fullInput as AnnotationPageNormalized;
        return getBestThumbnailAtSize(annotationPage.items[0], request, dereference, candidates, dimensions);
      }

      case 'Choice': {
        const choice = fullInput;
        if (!choice.items || choice.items[0]) {
          return await thumbnailNotFound();
        }
        // @todo this could also be configuration, just choosing the first choice.
        return getBestThumbnailAtSize(choice.items[0] as any, request, dereference, candidates, dimensions);
      }
      case 'Collection': {
        // This one is tricky, as the manifests may not have been loaded. But we will give it a shot.
        const collection = fullInput as CollectionNormalized;
        const firstManifest = collection.items[0];
        if (!firstManifest) {
          return await thumbnailNotFound();
        }
        return getBestThumbnailAtSize(firstManifest, request, dereference, candidates, dimensions);
      }

      case 'Manifest': {
        const manifest = fullInput as ManifestNormalized;
        const firstCanvas = manifest.items[0];
        if (!firstCanvas) {
          return await thumbnailNotFound();
        }
        return getBestThumbnailAtSize(firstCanvas, request, dereference, candidates, dimensions);
      }

      case 'SpecificResource':
      case 'Image':
      case 'Dataset':
      case 'Sound':
      case 'Text':
      case 'TextualBody':
      case 'Video':
        if (dimensions && !(fullInput as any).width) {
          (fullInput as any).width = dimensions.width;
          (fullInput as any).height = dimensions.height;
        }

        return loader.getThumbnailFromResource(fullInput as any, request, dereference, candidates);

      // Seems unlikely these would appear, but it would be an error..
      // case 'Service': // @todo could do something with vault.
      // case 'Range':
      // case 'AnnotationCollection':
      // case 'CanvasReference':
      // case 'ContentResource':
      //   return await thumbnailNotFound();
    }

    return await thumbnailNotFound();
  }

  return {
    getBestThumbnailAtSize,
  };
}
