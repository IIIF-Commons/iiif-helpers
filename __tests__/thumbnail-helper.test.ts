import { thumbnailFixtures } from '../fixtures';
import { expect, test } from 'vitest';
import { Vault } from '@iiif/vault';
import { createThumbnailHelper } from '../src/thumbnail';
import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { ManifestNormalized } from '@iiif/presentation-3';

describe('Thumbnail helper', function () {
  test.each(thumbnailFixtures as { label: string; description: string }[])(`Thumbnail - $label`, async (fixture) => {
    const vault = new Vault();
    const helper = createThumbnailHelper(vault);
    const manifestJson: any = JSON.parse(
      (await readFile(path.join(process.cwd(), 'fixtures/thumbnails', `${fixture.label}.json`))).toString()
    );
    const manifest = await vault.load<ManifestNormalized>(manifestJson.id || manifestJson['@id']);

    if (!manifest) {
      expect.fail(`Invalid manifest`);
    }

    const thumbnails = [];
    for (const canvas of manifest.items) {
      thumbnails.push(helper.getBestThumbnailAtSize(canvas, { width: 256, height: 256 }));
    }

    expect(await Promise.all(thumbnails)).toMatchSnapshot();
  });
});
