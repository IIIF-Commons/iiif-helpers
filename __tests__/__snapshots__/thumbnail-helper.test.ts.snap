// Vitest Snapshot v1

exports[`Thumbnail helper > Thumbnail - Canvas with no thumbnail property - ImageService (no vault) 1`] = `
[
  {
    "best": {
      "height": 512,
      "id": "https://iiif.io/api/image/2.1/example/reference/918ecd18c2592080851777620de9bcb5-fountain/full/512,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 512,
    },
    "fallback": [],
    "log": [],
  },
  {
    "best": {
      "height": 512,
      "id": "https://iiif.io/api/image/2.1/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/512,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 512,
    },
    "fallback": [],
    "log": [],
  },
]
`;

exports[`Thumbnail helper > Thumbnail - Canvas with no thumbnail property - ImageService 1`] = `
[
  {
    "best": {
      "height": 512,
      "id": "https://iiif.io/api/image/2.1/example/reference/918ecd18c2592080851777620de9bcb5-fountain/full/512,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 512,
    },
    "fallback": [],
    "log": [],
  },
  {
    "best": {
      "height": 512,
      "id": "https://iiif.io/api/image/2.1/example/reference/918ecd18c2592080851777620de9bcb5-gottingen/full/512,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 512,
    },
    "fallback": [],
    "log": [],
  },
]
`;

exports[`Thumbnail helper > Thumbnail - Canvas with no thumbnail property - level0 image (no vault) 1`] = `
[
  {
    "best": {
      "height": 189,
      "id": "https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/no_thum_prop/v2/level0/tractor/full/252,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 252,
    },
    "fallback": [],
    "log": [],
  },
  {
    "best": {
      "height": 189,
      "id": "https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/no_thum_prop/v2/level0/van/full/252,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 252,
    },
    "fallback": [],
    "log": [],
  },
]
`;

exports[`Thumbnail helper > Thumbnail - Canvas with no thumbnail property - level0 image 1`] = `
[
  {
    "best": {
      "height": 189,
      "id": "https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/no_thum_prop/v2/level0/tractor/full/252,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 252,
    },
    "fallback": [],
    "log": [],
  },
  {
    "best": {
      "height": 189,
      "id": "https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/no_thum_prop/v2/level0/van/full/252,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 252,
    },
    "fallback": [],
    "log": [],
  },
]
`;

exports[`Thumbnail helper > Thumbnail - Canvas with no thumbnail property - simple image (no vault) 1`] = `
[
  {
    "best": {
      "height": 4032,
      "id": "https://fixtures.iiif.io/images/Glen/photos/fountain.jpg",
      "type": "fixed",
      "unsafe": true,
      "width": 3024,
    },
    "fallback": [],
    "log": [],
  },
  {
    "best": undefined,
    "fallback": [],
    "log": [],
  },
]
`;

exports[`Thumbnail helper > Thumbnail - Canvas with no thumbnail property - simple image 1`] = `
[
  {
    "best": {
      "height": 4032,
      "id": "https://fixtures.iiif.io/images/Glen/photos/fountain.jpg",
      "type": "fixed",
      "unsafe": true,
      "width": 3024,
    },
    "fallback": [],
    "log": [],
  },
  {
    "best": undefined,
    "fallback": [],
    "log": [],
  },
]
`;

exports[`Thumbnail helper > Thumbnail - Canvases with thumbnail property which is an image URL (no vault) 1`] = `
[
  {
    "best": {
      "height": 512,
      "id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-4_titlepage1_verso/full/512,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 512,
    },
    "fallback": [
      {
        "height": 7230,
        "id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-4_titlepage1_verso/full/full/0/default.jpg",
        "type": "fixed",
        "unsafe": true,
        "width": 5428,
      },
    ],
    "log": [],
  },
  {
    "best": {
      "height": 512,
      "id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-6_titlepage2_verso/full/512,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 512,
    },
    "fallback": [
      {
        "height": 7230,
        "id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-6_titlepage2_verso/full/full/0/default.jpg",
        "type": "fixed",
        "unsafe": true,
        "width": 5428,
      },
    ],
    "log": [],
  },
]
`;

exports[`Thumbnail helper > Thumbnail - Canvases with thumbnail property which is an image URL 1`] = `
[
  {
    "best": {
      "height": 512,
      "id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-4_titlepage1_verso/full/512,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 512,
    },
    "fallback": [
      {
        "height": 7230,
        "id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-4_titlepage1_verso/full/full/0/default.jpg",
        "type": "fixed",
        "unsafe": true,
        "width": 5428,
      },
    ],
    "log": [],
  },
  {
    "best": {
      "height": 512,
      "id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-6_titlepage2_verso/full/512,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 512,
    },
    "fallback": [
      {
        "height": 7230,
        "id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-6_titlepage2_verso/full/full/0/default.jpg",
        "type": "fixed",
        "unsafe": true,
        "width": 5428,
      },
    ],
    "log": [],
  },
]
`;

exports[`Thumbnail helper > Thumbnail - Canvases with thumbnail property which is an image URL with dimensions (no vault) 1`] = `
[
  {
    "best": {
      "height": 512,
      "id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-4_titlepage1_verso/full/512,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 512,
    },
    "fallback": [
      {
        "height": 7230,
        "id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-4_titlepage1_verso/full/full/0/default.jpg",
        "type": "fixed",
        "unsafe": true,
        "width": 5428,
      },
    ],
    "log": [],
  },
  {
    "best": {
      "height": 512,
      "id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-6_titlepage2_verso/full/512,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 512,
    },
    "fallback": [
      {
        "height": 7230,
        "id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-6_titlepage2_verso/full/full/0/default.jpg",
        "type": "fixed",
        "unsafe": true,
        "width": 5428,
      },
    ],
    "log": [],
  },
]
`;

exports[`Thumbnail helper > Thumbnail - Canvases with thumbnail property which is an image URL with dimensions 1`] = `
[
  {
    "best": {
      "height": 512,
      "id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-4_titlepage1_verso/full/512,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 512,
    },
    "fallback": [
      {
        "height": 7230,
        "id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-4_titlepage1_verso/full/full/0/default.jpg",
        "type": "fixed",
        "unsafe": true,
        "width": 5428,
      },
    ],
    "log": [],
  },
  {
    "best": {
      "height": 512,
      "id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-6_titlepage2_verso/full/512,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 512,
    },
    "fallback": [
      {
        "height": 7230,
        "id": "https://iiif.io/api/image/2.1/example/reference/15f769d62ca9a3a2deca390efed75d73-6_titlepage2_verso/full/full/0/default.jpg",
        "type": "fixed",
        "unsafe": true,
        "width": 5428,
      },
    ],
    "log": [],
  },
]
`;

exports[`Thumbnail helper > Thumbnail - Canvases with thumbnail property with level0 image service (Most optimised option) (no vault) 1`] = `
[
  {
    "best": {
      "height": 452,
      "id": "https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/thum_prop/v2/canvas_thum_level0/titlepage1_verso/full/340,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 340,
    },
    "fallback": [],
    "log": [],
  },
  {
    "best": {
      "height": 452,
      "id": "https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/thum_prop/v2/canvas_thum_level0/titlepage2_verso/full/340,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 340,
    },
    "fallback": [],
    "log": [],
  },
]
`;

exports[`Thumbnail helper > Thumbnail - Canvases with thumbnail property with level0 image service (Most optimised option) 1`] = `
[
  {
    "best": {
      "height": 452,
      "id": "https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/thum_prop/v2/canvas_thum_level0/titlepage1_verso/full/340,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 340,
    },
    "fallback": [],
    "log": [],
  },
  {
    "best": {
      "height": 452,
      "id": "https://iiif-commons.github.io/fixtures/examples/thumbnail/canvas/thum_prop/v2/canvas_thum_level0/titlepage2_verso/full/340,/0/default.jpg",
      "type": "fixed",
      "unsafe": false,
      "width": 340,
    },
    "fallback": [],
    "log": [],
  },
]
`;
